File: java/src/main/java/com/meteroid/api/Plan.java
// this file is @generated
package com.meteroid.api;

import com.meteroid.MeteroidHttpClient;
import com.meteroid.Utils;
import com.meteroid.exceptions.ApiException;
import com.meteroid.models.PlanListResponse;

import okhttp3.HttpUrl;

import java.io.IOException;

public class Plan {
    private final MeteroidHttpClient client;

    public Plan(MeteroidHttpClient client) {
        this.client = client;
    }

    /** List plans with optional filtering by product family. */
    public PlanListResponse listPlans() throws IOException, ApiException {

        return this.listPlans(new PlanListPlansOptions());
    }

    /** List plans with optional filtering by product family. */
    public PlanListResponse listPlans(final PlanListPlansOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/plans");
        if (options.productFamilyId != null) {
            url.addQueryParameter(
                    "product_family_id", Utils.serializeQueryParam(options.productFamilyId));
        }
        if (options.page != null) {
            url.addQueryParameter("page", Utils.serializeQueryParam(options.page));
        }
        if (options.perPage != null) {
            url.addQueryParameter("per_page", Utils.serializeQueryParam(options.perPage));
        }
        return this.client.executeRequest("GET", url.build(), null, null, PlanListResponse.class);
    }

    /** Retrieve the details of a specific plan */
    public Plan getPlanDetails(final String planId) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client.newUrlBuilder().encodedPath(String.format("/api/v1/plans/%s", planId));
        return this.client.executeRequest("GET", url.build(), null, null, Plan.class);
    }
}

File: rust/src/api/plan.rs
// this file is @generated
#[allow(unused_imports)]
use crate::{error::Result, models::*, Configuration};

#[derive(Default)]
pub struct PlanListPlansOptions {
    pub product_family_id: Option<ProductFamilyId>,

    /// Page number (0-indexed)
    pub page: Option<i32>,

    /// Number of items per page
    pub per_page: Option<i32>,
}

pub struct Plan<'a> {
    cfg: &'a Configuration,
}

impl<'a> Plan<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self { cfg }
    }

    /// List plans with optional filtering by product family.
    pub async fn list_plans(
        &self,
        options: Option<PlanListPlansOptions>,
    ) -> Result<crate::models::PlanListResponse> {
        let PlanListPlansOptions {
            product_family_id,
            page,
            per_page,
        } = options.unwrap_or_default();

        crate::request::Request::new(http1::Method::GET, "/api/v1/plans")
            .with_optional_query_param("product_family_id", product_family_id)
            .with_optional_query_param("page", page)
            .with_optional_query_param("per_page", per_page)
            .execute(self.cfg)
            .await
    }

    /// Retrieve the details of a specific plan
    pub async fn get_plan_details(&self, plan_id: String) -> Result<crate::models::Plan> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/plans/{plan_id}")
            .with_path_param("plan_id", plan_id)
            .execute(self.cfg)
            .await
    }
}
